File: lint-arwaky/src-python/taxonomy/watch_service_error.py
"""watch_service_error — Watch service domain error types."""

from pydantic import BaseModel, ConfigDict

from .error_code_vo import ErrorCode
from .error_value_vo import (
    ErrorMessage,
    Cause,
)
from .file_path_vo import FilePath


class WatchServiceError(BaseModel):
    """General failure during file system watching."""

    model_config = ConfigDict(frozen=True)

    path: FilePath | None = None
    message: ErrorMessage
    error_code: ErrorCode | None = None
    cause: Cause | None = None

    def __str__(self):
        target = f" on {self.path}" if self.path else ""
        code = f" [{self.error_code}]" if self.error_code else ""
        return f"Watch Error{target}{code}: {self.message}"


class WatchSubscriptionError(WatchServiceError):
    """Failed to subscribe to path changes (e.g., too many watchers)."""
    pass


class WatchEventError(WatchServiceError):
    """Failed to process a specific file event."""
    pass

File: lint-arwaky/src-rust/taxonomy/watch_service_error.rs
use serde::{Serialize, Deserialize};
use super::*;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct WatchServiceError {
    #[serde(default)]
    pub path: Option<FilePath>,
    pub message: ErrorMessage,
    #[serde(default)]
    pub error_code: Option<ErrorCode>,
    #[serde(default)]
    pub cause: Option<Cause>,
}

impl WatchServiceError {
    pub fn new(message: ErrorMessage) -> Self {
        Self { path: None, message, error_code: None, cause: None }
    }
}

impl std::fmt::Display for WatchServiceError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let target = self.path.as_ref().map(|p| format!(" on {}", p)).unwrap_or_default();
        let code = self.error_code.as_ref().map(|c| format!(" [{}]", c)).unwrap_or_default();
        write!(f, "Watch Error{}{}: {}", target, code, self.message)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct WatchSubscriptionError {
    #[serde(flatten)]
    pub base: WatchServiceError,
}

impl WatchSubscriptionError {
    pub fn new(message: ErrorMessage) -> Self {
        Self { base: WatchServiceError::new(message) }
    }
}

impl std::fmt::Display for WatchSubscriptionError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.base)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct WatchEventError {
    #[serde(flatten)]
    pub base: WatchServiceError,
}

impl WatchEventError {
    pub fn new(message: ErrorMessage) -> Self {
        Self { base: WatchServiceError::new(message) }
    }
}

impl std::fmt::Display for WatchEventError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.base)
    }
}
